File: docs/beginner/module-05-data-sources-indexers/code-samples/csharp/05_IndexerScheduling.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Search.Documents.Indexes;
using Azure.Search.Documents.Indexes.Models;

namespace AzureSearchSamples.DataSourcesIndexers
{
    /// <summary>
    /// Indexer Scheduling & Automation Example
    /// 
    /// This example demonstrates how to configure indexer schedules and automation patterns
    /// for Azure AI Search indexers.
    /// </summary>
    public class IndexerSchedulingExample
    {
        private readonly SearchIndexerClient _indexerClient;
        private readonly string _searchEndpoint;
        private readonly string _apiKey;

        public IndexerSchedulingExample(string searchEndpoint, string apiKey)
        {
            _searchEndpoint = searchEndpoint ?? throw new ArgumentNullException(nameof(searchEndpoint));
            _apiKey = apiKey ?? throw new ArgumentNullException(nameof(apiKey));

            var credential = new AzureKeyCredential(_apiKey);
            _indexerClient = new SearchIndexerClient(new Uri(_searchEndpoint), credential);
        }

        public async Task RunAsync()
        {
            Console.WriteLine("🚀 Indexer Scheduling & Automation Example");
            Console.WriteLine("=" + new string('=', 49));

            try
            {
                // Demonstrate scheduling patterns
                var patterns = DemonstrateSchedulingPatterns();

                // Create scheduling examples
                var scheduleExamples = CreateScheduledIndexerExamples();

                // Show best practices
                DemonstrateSchedulingBestPractices();

                // Monitor existing indexers (if any)
                await MonitorScheduledIndexersAsync();

                // Show optimization strategies
                DemonstrateOptimizationStrategies();

                // Troubleshooting guide
                TroubleshootSchedulingIssues();

                Console.WriteLine("\n✅ Indexer scheduling example completed successfully!");
                Console.WriteLine("\nKey takeaways:");
                Console.WriteLine("- Balance frequency with cost - more frequent updates cost more resources");
                Console.WriteLine("- Use change detection - minimize processing overhead with incremental updates");
                Console.WriteLine("- Schedule during off-peak hours - optimize performance and resource usage");
                Console.WriteLine("- Monitor execution patterns - track performance trends and adjust accordingly");
                Console.WriteLine("- Implement proper error handling - ensure reliable automated operations");
                Console.WriteLine("- Consider business requirements - align technical scheduling with business needs");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"\n❌ Example failed: {ex.Message}");
                throw;
            }
        }

        private List<dynamic> DemonstrateSchedulingPatterns()
        {
            Console.WriteLine("\n📅 Common Indexer Scheduling Patterns");
            Console.WriteLine("=" + new string('=', 39));

            var patterns = new[]
            {
                new
                {
                    Name = "High-Frequency Updates",
                    Interval = TimeSpan.FromMinutes(15),
                    UseCase = "Real-time data feeds, news, social media",
                    Cost = "High",
                    DataFreshness = "Very High",
                    ExampleStart = DateTime.UtcNow.Date.AddHours(DateTime.UtcNow.Hour)
                },
                new
                {
                    Name = "Hourly Updates",
                    Interval = TimeSpan.FromHours(1),
                    UseCase = "E-commerce catalogs, inventory systems",
                    Cost = "Medium-High",
                    DataFreshness = "High",
                    ExampleStart = DateTime.UtcNow.Date.AddHours(DateTime.UtcNow.Hour)
                },
                new
                {
                    Name = "Daily Updates",
                    Interval = TimeSpan.FromDays(1),
                    UseCase = "Document repositories, CRM data",
                    Cost = "Medium",
                    DataFreshness = "Medium",
                    ExampleStart = DateTime.UtcNow.Date.AddHours(2)
                },
                new
                {
                    Name = "Weekly Updates",
                    Interval = TimeSpan.FromDays(7),
                    UseCase = "Archive data, reference materials",
                    Cost = "Low",
                    DataFreshness = "Low",
                    ExampleStart = DateTime.UtcNow.Date.AddHours(1)
                }
            };

            foreach (var pattern in patterns)
            {
                Console.WriteLine($"\n🔄 {pattern.Name}");
                Console.WriteLine($"   Interval: {pattern.Interval}");
                Console.WriteLine($"   Use Case: {pattern.UseCase}");
                Console.WriteLine($"   Cost Impact: {pattern.Cost}");
                Console.WriteLine($"   Data Freshness: {pattern.DataFreshness}");
                Console.WriteLine($"   Example Start Time: {pattern.ExampleStart}");

                // Calculate next few runs
                var nextRuns = new List<DateTime>();
                var currentTime = pattern.ExampleStart;
                for (int i = 0; i < 3; i++)
                {
                    nextRuns.Add(currentTime.Add(TimeSpan.FromTicks(pattern.Interval.Ticks * i)));
                }

                Console.WriteLine($"   Next 3 runs: {string.Join(", ", nextRuns.Select(t => t.ToString("HH:mm")))}");
            }

            return patterns.Cast<dynamic>().ToList();
        }

        private Dictionary<string, IndexingSchedule> CreateScheduledIndexerExamples()
        {
            Console.WriteLine("\n⚙️ Creating Scheduled Indexer Examples");
            Console.WriteLine("=" + new string('=', 39));

            var examples = new Dictionary<string, IndexingSchedule>();

            // Example 1: E-commerce Product Catalog
            Console.WriteLine("\n🛒 E-commerce Product Catalog Indexer:");
            var ecommerceSchedule = new IndexingSchedule(
                interval: TimeSpan.FromHours(2),
                startTime: DateTimeOffset.UtcNow.Date.AddHours(8));

            Console.WriteLine($"   Interval: {ecommerceSchedule.Interval}");
            Console.WriteLine($"   Start Time: {ecommerceSchedule.StartTime} (8 AM)");
            Console.WriteLine("   Rationale: Frequent updates for inventory and pricing changes");
            Console.WriteLine("   Business Impact: High - affects customer experience and sales");

            examples.Add("E-commerce", ecommerceSchedule);

            // Example 2: Document Management System
            Console.WriteLine("\n📄 Document Management System Indexer:");
            var documentSchedule = new IndexingSchedule(
                interval: TimeSpan.FromDays(1),
                startTime: DateTimeOffset.UtcNow.Date.AddHours(2).AddMinutes(30));

            Console.WriteLine($"   Interval: {documentSchedule.Interval}");
            Console.WriteLine($"   Start Time: {documentSchedule.StartTime} (2:30 AM)");
            Console.WriteLine("   Rationale: Documents change less frequently, off-hours processing");
            Console.WriteLine("   Business Impact: Medium - affects search accuracy but not critical");

            examples.Add("Documents", documentSchedule);

            // Example 3: News and Media Content
            Console.WriteLine("\n📰 News and Media Content Indexer:");
            var newsSchedule = new IndexingSchedule(
                interval: TimeSpan.FromMinutes(15),
                startTime: DateTimeOffset.UtcNow.Date.AddHours(DateTime.UtcNow.Hour));

            Console.WriteLine($"   Interval: {newsSchedule.Interval}");
            Console.WriteLine($"   Start Time: {newsSchedule.StartTime}");
            Console.WriteLine("   Rationale: Breaking news requires immediate indexing");
            Console.WriteLine("   Business Impact: Critical - affects content relevance and user engagement");

            examples.Add("News", newsSchedule);

            // Example 4: Reference Data (Low Priority)
            Console.WriteLine("\n📚 Reference Data Indexer:");
            var referenceSchedule = new IndexingSchedule(
                interval: TimeSpan.FromDays(7),
                startTime: DateTimeOffset.UtcNow.Date.AddHours(1));

            Console.WriteLine($"   Interval: {referenceSchedule.Interval}");
            Console.WriteLine($"   Start Time: {referenceSchedule.StartTime} (1 AM Sunday)");
            Console.WriteLine("   Rationale: Reference data changes infrequently");
            Console.WriteLine("   Business Impact: Low - static content with minimal updates");

            examples.Add("Reference", referenceSchedule);

            return examples;
        }

        private void DemonstrateSchedulingBestPractices()
        {
            Console.WriteLine("\n💡 Indexer Scheduling Best Practices");
            Console.WriteLine("=" + new string('=', 34));

            var bestPractices = new[]
            {
                new
                {
                    Category = "Timing Optimization",
                    Practices = new[]
                    {
                        "Schedule during off-peak hours (2-4 AM)",
                        "Avoid overlapping with backup windows",
                        "Consider time zones for global applications",
                        "Stagger multiple indexers to avoid resource conflicts"
                    }
                },
                new
                {
                    Category = "Performance Considerations",
                    Practices = new[]
                    {
                        "Use change detection to minimize processing",
                        "Adjust batch sizes based on data volume",
                        "Monitor execution duration trends",
                        "Set appropriate timeout values"
                    }
                },
                new
                {
                    Category = "Cost Management",
                    Practices = new[]
                    {
                        "Balance frequency with business requirements",
                        "Use incremental indexing when possible",
                        "Monitor search unit consumption",
                        "Consider data source throttling limits"
                    }
                },
                new
                {
                    Category = "Reliability & Monitoring",
                    Practices = new[]
                    {
                        "Set up alerting for failed executions",
                        "Monitor execution history regularly",
                        "Implement retry logic for transient failures",
                        "Document schedule rationale and dependencies"
                    }
                }
            };

            foreach (var category in bestPractices)
            {
                Console.WriteLine($"\n🎯 {category.Category}");
                foreach (var practice in category.Practices)
                    Console.WriteLine($"   • {practice}");
            }

            // Create a scheduling decision matrix
            Console.WriteLine($"\n📊 Scheduling Decision Matrix:");
            Console.WriteLine("   Data Volatility | Business Criticality | Data Volume | Recommended Frequency");
            Console.WriteLine("   Low             | Low                  | Small       | Weekly");
            Console.WriteLine("   Medium          | Medium               | Medium      | Daily");
            Console.WriteLine("   High            | High                 | Large       | Hourly");
            Console.WriteLine("   Very High       | Critical             | Very Large  | 15-30 min");
        }

        private async Task MonitorScheduledIndexersAsync()
        {
            Console.WriteLine("\n📊 Monitoring Scheduled Indexers");
            Console.WriteLine("=" + new string('=', 29));

            try
            {
                // Get all indexers
                var indexers = new List<SearchIndexer>();
                await foreach (var indexer in _indexerClient.GetIndexersAsync())
                {
                    indexers.Add(indexer);
                }

                if (!indexers.Any())
                {
                    Console.WriteLine("⚠️ No indexers found in the search service");
                    return;
                }

                var monitoringData = new List<dynamic>();

                foreach (var indexer in indexers)
                {
                    Console.WriteLine($"\n🔍 Analyzing indexer: {indexer.Name}");

                    try
                    {
                        var status = await _indexerClient.GetIndexerStatusAsync(indexer.Name);

                        // Collect basic info
                        var indexerInfo = new
                        {
                            Name = indexer.Name,
                            Status = status.Value.Status.ToString(),
                            HasSchedule = indexer.Schedule != null,
                            ScheduleInterval = indexer.Schedule?.Interval.ToString() ?? "Manual",
                            LastResultStatus = status.Value.LastResult?.Status.ToString() ?? "None",
                            ExecutionCount = status.Value.ExecutionHistory?.Count ?? 0
                        };

                        var itemsProcessed = 0;
                        var itemsFailed = 0;
                        var executionDuration = 0.0;

                        if (status.Value.LastResult != null)
                        {
                            var result = status.Value.LastResult;
                            itemsProcessed = result.ItemCount;
                            itemsFailed = result.FailedItemCount;

                            if (result.EndTime.HasValue && result.StartTime.HasValue)
                            {
                                executionDuration = (result.EndTime.Value - result.StartTime.Value).TotalSeconds;
                            }
                        }

                        // Display indexer details
                        Console.WriteLine($"   Status: {indexerInfo.Status}");
                        Console.WriteLine($"   Schedule: {indexerInfo.ScheduleInterval}");
                        Console.WriteLine($"   Last Result: {indexerInfo.LastResultStatus}");
                        Console.WriteLine($"   Items Processed: {itemsProcessed}");
                        Console.WriteLine($"   Execution Duration: {executionDuration:F1}s");
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"   ❌ Error getting status: {ex.Message}");
                        continue;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Error monitoring indexers: {ex.Message}");
            }
        }

        private void DemonstrateOptimizationStrategies()
        {
            Console.WriteLine("\n🚀 Advanced Scheduling Optimization Strategies");
            Console.WriteLine("=" + new string('=', 44));

            var strategies = new[]
            {
                new
                {
                    Name = "Adaptive Scheduling",
                    Description = "Adjust frequency based on data change patterns",
                    Implementation = new[]
                    {
                        "Monitor change detection results",
                        "Reduce frequency when few changes detected",
                        "Increase frequency during high-activity periods",
                        "Use Azure Logic Apps for dynamic scheduling"
                    },
                    Benefits = "Optimal balance of freshness and cost"
                },
                new
                {
                    Name = "Tiered Scheduling",
                    Description = "Different schedules for different data priorities",
                    Implementation = new[]
                    {
                        "Critical data: High frequency (15-30 min)",
                        "Important data: Medium frequency (hourly)",
                        "Archive data: Low frequency (daily/weekly)",
                        "Use separate indexers for each tier"
                    },
                    Benefits = "Resource optimization and cost control"
                },
                new
                {
                    Name = "Event-Driven Indexing",
                    Description = "Trigger indexing based on data source events",
                    Implementation = new[]
                    {
                        "Use Azure Event Grid for blob storage changes",
                        "Implement webhook triggers for database changes",
                        "Combine with scheduled baseline runs",
                        "Use Azure Functions for event processing"
                    },
                    Benefits = "Near real-time updates with minimal overhead"
                },
                new
                {
                    Name = "Load Balancing",
                    Description = "Distribute indexer execution across time",
                    Implementation = new[]
                    {
                        "Stagger start times for multiple indexers",
                        "Use different intervals to avoid conflicts",
                        "Monitor search unit utilization",
                        "Implement queue-based execution"
                    },
                    Benefits = "Improved performance and resource utilization"
                }
            };

            foreach (var strategy in strategies)
            {
                Console.WriteLine($"\n🎯 {strategy.Name}");
                Console.WriteLine($"   Description: {strategy.Description}");
                Console.WriteLine($"   Benefits: {strategy.Benefits}");
                Console.WriteLine("   Implementation:");
                foreach (var step in strategy.Implementation)
                    Console.WriteLine($"     • {step}");
            }
        }

        private void TroubleshootSchedulingIssues()
        {
            Console.WriteLine("\n🔧 Common Scheduling Issues and Solutions");
            Console.WriteLine("=" + new string('=', 39));

            var issues = new[]
            {
                new
                {
                    Issue = "Indexer Not Running on Schedule",
                    Symptoms = new[]
                    {
                        "No recent execution history",
                        "Status shows as idle",
                        "Data not updating as expected"
                    },
                    Causes = new[]
                    {
                        "Schedule not properly configured",
                        "Indexer disabled or in error state",
                        "Service quota limits reached",
                        "Data source connection issues"
                    },
                    Solutions = new[]
                    {
                        "Verify schedule configuration",
                        "Check indexer status and enable if needed",
                        "Review service limits and usage",
                        "Test data source connectivity"
                    }
                },
                new
                {
                    Issue = "Execution Timeouts",
                    Symptoms = new[]
                    {
                        "Indexer stops mid-execution",
                        "Partial data processing",
                        "Timeout errors in execution history"
                    },
                    Causes = new[]
                    {
                        "Large batch sizes",
                        "Complex data transformations",
                        "Slow data source responses",
                        "Network connectivity issues"
                    },
                    Solutions = new[]
                    {
                        "Reduce batch size",
                        "Optimize field mappings",
                        "Implement data source caching",
                        "Increase timeout settings"
                    }
                },
                new
                {
                    Issue = "Resource Conflicts",
                    Symptoms = new[]
                    {
                        "Multiple indexers failing simultaneously",
                        "Performance degradation",
                        "Search unit exhaustion"
                    },
                    Causes = new[]
                    {
                        "Overlapping execution schedules",
                        "Insufficient search units",
                        "Concurrent indexer limits exceeded"
                    },
                    Solutions = new[]
                    {
                        "Stagger indexer schedules",
                        "Upgrade service tier",
                        "Implement execution queuing"
                    }
                }
            };

            foreach (var issue in issues)
            {
                Console.WriteLine($"\n❌ {issue.Issue}");
                Console.WriteLine("   Symptoms:");
                foreach (var symptom in issue.Symptoms)
                    Console.WriteLine($"     • {symptom}");
                Console.WriteLine("   Common Causes:");
                foreach (var cause in issue.Causes)
                    Console.WriteLine($"     • {cause}");
                Console.WriteLine("   Solutions:");
                foreach (var solution in issue.Solutions)
                    Console.WriteLine($"     ✅ {solution}");
            }

            // Diagnostic checklist
            Console.WriteLine($"\n📋 Scheduling Diagnostic Checklist:");
            var checklist = new[]
            {
                "✓ Verify indexer schedule configuration",
                "✓ Check indexer status and last execution",
                "✓ Review execution history for patterns",
                "✓ Monitor search unit consumption",
                "✓ Test data source connectivity",
                "✓ Validate change detection settings",
                "✓ Check for service limit violations",
                "✓ Review error logs and messages",
                "✓ Verify field mapping configurations",
                "✓ Test manual indexer execution"
            };

            foreach (var item in checklist)
                Console.WriteLine($"   {item}");
        }
    }

    public class Program
    {
        public static async Task Main(string[] args)
        {
            var searchEndpoint = Environment.GetEnvironmentVariable("SEARCH_ENDPOINT") ?? "https://your-search-service.search.windows.net";
            var apiKey = Environment.GetEnvironmentVariable("SEARCH_API_KEY") ?? "your-admin-api-key";

            var example = new IndexerSchedulingExample(searchEndpoint, apiKey);

            try
            {
                await example.RunAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Application failed: {ex.Message}");
                Environment.Exit(1);
            }

            Console.WriteLine("\nPress any key to exit...");
            Console.ReadKey();
        }
    }
}

File: docs/beginner/module-05-data-sources-indexers/code-samples/python/05_indexer_scheduling.py
#!/usr/bin/env python3
"""
Indexer Scheduling & Automation Example

This script demonstrates how to configure indexer schedules and automation patterns
for Azure AI Search indexers.

Prerequisites:
- Azure AI Search service
- Existing indexers to schedule
- Admin API key or managed identity
- Required Python packages installed
"""

import os
import time
from datetime import datetime, timedelta
from dotenv import load_dotenv

from azure.search.documents.indexes import SearchIndexerClient
from azure.search.documents.indexes.models import (
    IndexingSchedule, SearchIndexer
)
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError

# Load environment variables
load_dotenv()

# Configuration
SEARCH_ENDPOINT = os.getenv('SEARCH_ENDPOINT')
SEARCH_API_KEY = os.getenv('SEARCH_API_KEY')

def validate_configuration():
    """Validate that required configuration is present."""
    if not all([SEARCH_ENDPOINT, SEARCH_API_KEY]):
        raise ValueError("Missing required search service configuration.")
    
    print("✅ Configuration validated")
    print(f"📍 Search Endpoint: {SEARCH_ENDPOINT}")

def demonstrate_scheduling_patterns():
    """Demonstrate different indexer scheduling patterns."""
    print("\n📅 Common Indexer Scheduling Patterns")
    print("=" * 40)
    
    patterns = [
        {
            'name': 'High-Frequency Updates',
            'interval': timedelta(minutes=15),
            'use_case': 'Real-time data feeds, news, social media',
            'cost': 'High',
            'data_freshness': 'Very High',
            'example_start': datetime.now().replace(second=0, microsecond=0)
        },
        {
            'name': 'Hourly Updates',
            'interval': timedelta(hours=1),
            'use_case': 'E-commerce catalogs, inventory systems',
            'cost': 'Medium-High',
            'data_freshness': 'High',
            'example_start': datetime.now().replace(minute=0, second=0, microsecond=0)
        },
        {
            'name': 'Daily Updates',
            'interval': timedelta(days=1),
            'use_case': 'Document repositories, CRM data',
            'cost': 'Medium',
            'data_freshness': 'Medium',
            'example_start': datetime.now().replace(hour=2, minute=0, second=0, microsecond=0)
        },
        {
            'name': 'Weekly Updates',
            'interval': timedelta(days=7),
            'use_case': 'Archive data, reference materials',
            'cost': 'Low',
            'data_freshness': 'Low',
            'example_start': datetime.now().replace(hour=1, minute=0, second=0, microsecond=0)
        }
    ]
    
    for pattern in patterns:
        print(f"\n🔄 {pattern['name']}")
        print(f"   Interval: {pattern['interval']}")
        print(f"   Use Case: {pattern['use_case']}")
        print(f"   Cost Impact: {pattern['cost']}")
        print(f"   Data Freshness: {pattern['data_freshness']}")
        print(f"   Example Start Time: {pattern['example_start']}")
        
        # Calculate next few runs
        next_runs = []
        current_time = pattern['example_start']
        for i in range(3):
            next_runs.append(current_time + (pattern['interval'] * i))
        
        print(f"   Next 3 runs: {', '.join([t.strftime('%H:%M') for t in next_runs])}")
    
    return patterns

def create_scheduled_indexer_examples():
    """Create example indexer schedules for different scenarios."""
    print("\n⚙️ Creating Scheduled Indexer Examples")
    print("=" * 40)
    
    examples = []
    
    # Example 1: E-commerce Product Catalog
    print("\n🛒 E-commerce Product Catalog Indexer:")
    ecommerce_schedule = IndexingSchedule(
        interval=timedelta(hours=2),  # Every 2 hours during business hours
        start_time=datetime.now().replace(hour=8, minute=0, second=0, microsecond=0)
    )
    
    print(f"   Interval: {ecommerce_schedule.interval}")
    print(f"   Start Time: {ecommerce_schedule.start_time} (8 AM)")
    print("   Rationale: Frequent updates for inventory and pricing changes")
    print("   Business Impact: High - affects customer experience and sales")
    
    examples.append(('E-commerce', ecommerce_schedule))
    
    # Example 2: Document Management System
    print("\n📄 Document Management System Indexer:")
    document_schedule = IndexingSchedule(
        interval=timedelta(days=1),
        start_time=datetime.now().replace(hour=2, minute=30, second=0, microsecond=0)
    )
    
    print(f"   Interval: {document_schedule.interval}")
    print(f"   Start Time: {document_schedule.start_time} (2:30 AM)")
    print("   Rationale: Documents change less frequently, off-hours processing")
    print("   Business Impact: Medium - affects search accuracy but not critical")
    
    examples.append(('Documents', document_schedule))
    
    # Example 3: News and Media Content
    print("\n📰 News and Media Content Indexer:")
    news_schedule = IndexingSchedule(
        interval=timedelta(minutes=15),
        start_time=datetime.now().replace(second=0, microsecond=0)
    )
    
    print(f"   Interval: {news_schedule.interval}")
    print(f"   Start Time: {news_schedule.start_time}")
    print("   Rationale: Breaking news requires immediate indexing")
    print("   Business Impact: Critical - affects content relevance and user engagement")
    
    examples.append(('News', news_schedule))
    
    # Example 4: Reference Data (Low Priority)
    print("\n📚 Reference Data Indexer:")
    reference_schedule = IndexingSchedule(
        interval=timedelta(days=7),
        start_time=datetime.now().replace(hour=1, minute=0, second=0, microsecond=0)
    )
    
    print(f"   Interval: {reference_schedule.interval}")
    print(f"   Start Time: {reference_schedule.start_time} (1 AM Sunday)")
    print("   Rationale: Reference data changes infrequently")
    print("   Business Impact: Low - static content with minimal updates")
    
    examples.append(('Reference', reference_schedule))
    
    return examples

def demonstrate_scheduling_best_practices():
    """Demonstrate scheduling best practices and considerations."""
    print("\n💡 Indexer Scheduling Best Practices")
    print("=" * 35)
    
    best_practices = [
        {
            'category': 'Timing Optimization',
            'practices': [
                'Schedule during off-peak hours (2-4 AM)',
                'Avoid overlapping with backup windows',
                'Consider time zones for global applications',
                'Stagger multiple indexers to avoid resource conflicts'
            ]
        },
        {
            'category': 'Performance Considerations',
            'practices': [
                'Use change detection to minimize processing',
                'Adjust batch sizes based on data volume',
                'Monitor execution duration trends',
                'Set appropriate timeout values'
            ]
        },
        {
            'category': 'Cost Management',
            'practices': [
                'Balance frequency with business requirements',
                'Use incremental indexing when possible',
                'Monitor search unit consumption',
                'Consider data source throttling limits'
            ]
        },
        {
            'category': 'Reliability & Monitoring',
            'practices': [
                'Set up alerting for failed executions',
                'Monitor execution history regularly',
                'Implement retry logic for transient failures',
                'Document schedule rationale and dependencies'
            ]
        }
    ]
    
    for category in best_practices:
        print(f"\n🎯 {category['category']}")
        for practice in category['practices']:
            print(f"   • {practice}")
    
    # Create a scheduling decision matrix
    print(f"\n📊 Scheduling Decision Matrix:")
    print("   Data Volatility | Business Criticality | Data Volume | Recommended Frequency")
    print("   Low             | Low                  | Small       | Weekly")
    print("   Medium          | Medium               | Medium      | Daily")
    print("   High            | High                 | Large       | Hourly")
    print("   Very High       | Critical             | Very Large  | 15-30 min")

def apply_schedule_to_indexer(indexer_client, indexer_name, schedule):
    """Apply a schedule to an existing indexer."""
    print(f"\n⚙️ Applying schedule to indexer: {indexer_name}")
    
    try:
        # Get existing indexer
        indexer = indexer_client.get_indexer(indexer_name)
        
        # Update with schedule
        indexer.schedule = schedule
        
        # Update the indexer
        result = indexer_client.create_or_update_indexer(indexer)
        
        print(f"   ✅ Schedule applied successfully")
        print(f"   Interval: {result.schedule.interval}")
        print(f"   Start Time: {result.schedule.start_time}")
        
        return result
        
    except HttpResponseError as e:
        if e.status_code == 404:
            print(f"   ⚠️ Indexer '{indexer_name}' not found")
        else:
            print(f"   ❌ Error applying schedule: {e.message}")
        return None

def monitor_scheduled_indexers(indexer_client):
    """Monitor scheduled indexer performance and execution patterns."""
    print("\n📊 Monitoring Scheduled Indexers")
    print("=" * 30)
    
    try:
        # Get all indexers
        indexers = list(indexer_client.get_indexers())
        
        if not indexers:
            print("⚠️ No indexers found in the search service")
            return []
        
        monitoring_data = []
        
        for indexer in indexers:
            print(f"\n🔍 Analyzing indexer: {indexer.name}")
            
            try:
                status = indexer_client.get_indexer_status(indexer.name)
                
                # Collect basic info
                indexer_info = {
                    'name': indexer.name,
                    'status': status.status,
                    'has_schedule': indexer.schedule is not None,
                    'schedule_interval': str(indexer.schedule.interval) if indexer.schedule else 'Manual',
                    'last_result_status': status.last_result.status if status.last_result else 'None',
                    'execution_count': len(status.execution_history) if status.execution_history else 0
                }
                
                if status.last_result:
                    result = status.last_result
                    indexer_info.update({
                        'items_processed': result.item_count or 0,
                        'items_failed': result.failed_item_count or 0,
                        'last_execution': result.end_time or result.start_time,
                        'execution_duration': (result.end_time - result.start_time).total_seconds() if result.end_time and result.start_time else 0
                    })
                
                monitoring_data.append(indexer_info)
                
                # Display indexer details
                print(f"   Status: {indexer_info['status']}")
                print(f"   Schedule: {indexer_info['schedule_interval']}")
                print(f"   Last Result: {indexer_info['last_result_status']}")
                if 'items_processed' in indexer_info:
                    print(f"   Items Processed: {indexer_info['items_processed']}")
                    print(f"   Execution Duration: {indexer_info['execution_duration']:.1f}s")
                
            except Exception as e:
                print(f"   ❌ Error getting status: {str(e)}")
                continue
        
        return monitoring_data
        
    except Exception as e:
        print(f"❌ Error monitoring indexers: {str(e)}")
        return []

def demonstrate_optimization_strategies():
    """Demonstrate advanced scheduling optimization strategies."""
    print("\n🚀 Advanced Scheduling Optimization Strategies")
    print("=" * 45)
    
    strategies = [
        {
            'name': 'Adaptive Scheduling',
            'description': 'Adjust frequency based on data change patterns',
            'implementation': [
                'Monitor change detection results',
                'Reduce frequency when few changes detected',
                'Increase frequency during high-activity periods',
                'Use Azure Logic Apps for dynamic scheduling'
            ],
            'benefits': 'Optimal balance of freshness and cost'
        },
        {
            'name': 'Tiered Scheduling',
            'description': 'Different schedules for different data priorities',
            'implementation': [
                'Critical data: High frequency (15-30 min)',
                'Important data: Medium frequency (hourly)',
                'Archive data: Low frequency (daily/weekly)',
                'Use separate indexers for each tier'
            ],
            'benefits': 'Resource optimization and cost control'
        },
        {
            'name': 'Event-Driven Indexing',
            'description': 'Trigger indexing based on data source events',
            'implementation': [
                'Use Azure Event Grid for blob storage changes',
                'Implement webhook triggers for database changes',
                'Combine with scheduled baseline runs',
                'Use Azure Functions for event processing'
            ],
            'benefits': 'Near real-time updates with minimal overhead'
        },
        {
            'name': 'Load Balancing',
            'description': 'Distribute indexer execution across time',
            'implementation': [
                'Stagger start times for multiple indexers',
                'Use different intervals to avoid conflicts',
                'Monitor search unit utilization',
                'Implement queue-based execution'
            ],
            'benefits': 'Improved performance and resource utilization'
        }
    ]
    
    for strategy in strategies:
        print(f"\n🎯 {strategy['name']}")
        print(f"   Description: {strategy['description']}")
        print(f"   Benefits: {strategy['benefits']}")
        print("   Implementation:")
        for step in strategy['implementation']:
            print(f"     • {step}")

def troubleshoot_scheduling_issues():
    """Demonstrate troubleshooting techniques for scheduled indexers."""
    print("\n🔧 Common Scheduling Issues and Solutions")
    print("=" * 40)
    
    issues = [
        {
            'issue': 'Indexer Not Running on Schedule',
            'symptoms': [
                'No recent execution history',
                'Status shows as idle',
                'Data not updating as expected'
            ],
            'causes': [
                'Schedule not properly configured',
                'Indexer disabled or in error state',
                'Service quota limits reached',
                'Data source connection issues'
            ],
            'solutions': [
                'Verify schedule configuration',
                'Check indexer status and enable if needed',
                'Review service limits and usage',
                'Test data source connectivity'
            ]
        },
        {
            'issue': 'Execution Timeouts',
            'symptoms': [
                'Indexer stops mid-execution',
                'Partial data processing',
                'Timeout errors in execution history'
            ],
            'causes': [
                'Large batch sizes',
                'Complex data transformations',
                'Slow data source responses',
                'Network connectivity issues'
            ],
            'solutions': [
                'Reduce batch size',
                'Optimize field mappings',
                'Implement data source caching',
                'Increase timeout settings'
            ]
        },
        {
            'issue': 'Resource Conflicts',
            'symptoms': [
                'Multiple indexers failing simultaneously',
                'Performance degradation',
                'Search unit exhaustion'
            ],
            'causes': [
                'Overlapping execution schedules',
                'Insufficient search units',
                'Concurrent indexer limits exceeded'
            ],
            'solutions': [
                'Stagger indexer schedules',
                'Upgrade service tier',
                'Implement execution queuing'
            ]
        }
    ]
    
    for issue in issues:
        print(f"\n❌ {issue['issue']}")
        print("   Symptoms:")
        for symptom in issue['symptoms']:
            print(f"     • {symptom}")
        print("   Common Causes:")
        for cause in issue['causes']:
            print(f"     • {cause}")
        print("   Solutions:")
        for solution in issue['solutions']:
            print(f"     ✅ {solution}")
    
    # Diagnostic checklist
    print(f"\n📋 Scheduling Diagnostic Checklist:")
    checklist = [
        "✓ Verify indexer schedule configuration",
        "✓ Check indexer status and last execution",
        "✓ Review execution history for patterns",
        "✓ Monitor search unit consumption",
        "✓ Test data source connectivity",
        "✓ Validate change detection settings",
        "✓ Check for service limit violations",
        "✓ Review error logs and messages",
        "✓ Verify field mapping configurations",
        "✓ Test manual indexer execution"
    ]
    
    for item in checklist:
        print(f"   {item}")

def main():
    """Main execution function."""
    print("🚀 Indexer Scheduling & Automation Example")
    print("=" * 50)
    
    try:
        # Validate configuration
        validate_configuration()
        
        # Initialize client
        credential = AzureKeyCredential(SEARCH_API_KEY)
        indexer_client = SearchIndexerClient(SEARCH_ENDPOINT, credential)
        
        # Demonstrate scheduling patterns
        patterns = demonstrate_scheduling_patterns()
        
        # Create scheduling examples
        schedule_examples = create_scheduled_indexer_examples()
        
        # Show best practices
        demonstrate_scheduling_best_practices()
        
        # Monitor existing indexers
        monitoring_data = monitor_scheduled_indexers(indexer_client)
        
        # Show optimization strategies
        demonstrate_optimization_strategies()
        
        # Troubleshooting guide
        troubleshoot_scheduling_issues()
        
        print("\n✅ Indexer scheduling example completed successfully!")
        print("\nKey takeaways:")
        print("- Balance frequency with cost - more frequent updates cost more resources")
        print("- Use change detection - minimize processing overhead with incremental updates")
        print("- Schedule during off-peak hours - optimize performance and resource usage")
        print("- Monitor execution patterns - track performance trends and adjust accordingly")
        print("- Implement proper error handling - ensure reliable automated operations")
        print("- Consider business requirements - align technical scheduling with business needs")
        
    except Exception as e:
        print(f"\n❌ Example failed: {str(e)}")
        raise

if __name__ == "__main__":
    main()